Add functions for generating random unsigned integers

donny-dont opened the pull request:
#### ecc067cc894ba2088d6da3945e547491c2fdcdb5
<pre>
Add functions for generating random unsigned integers
<a href="https://bugs.webkit.org/show_bug.cgi?id=247446">https://bugs.webkit.org/show_bug.cgi?id=247446</a>

Reviewed by Yusuke Suzuki.

Add `cryptographicallyRandomUint32` and `cryptographicallyRandomUint64`
for generating unsigned integers. Replace uses of `randomNumber` that
were creating random unsigned integers. These were all just reversing
the operations used internally in `randomNumber`.

* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
* Source/WTF/wtf/RandomNumber.cpp:
* Source/WTF/wtf/RandomNumber.h:
* Source/WebCore/platform/network/FormDataBuilder.cpp:
* Source/WebCore/style/RuleSetBuilder.cpp:
* Source/WebKit/Platform/IPC/win/ConnectionWin.cpp:
* Source/WebKit/Platform/unix/SharedMemoryUnix.cpp:

Canonical link: <a href="https://commits.webkit.org/256293@main">https://commits.webkit.org/256293@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/36845a229b6e54426e1787e3842a011fa45a6034

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/95324 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/4604 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/28377 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/104917 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/165179 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/99311 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/4605 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/33323 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/87693 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/100800 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/100988 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/3358 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/81907 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/30436 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/85259 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/28377 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/73276 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/86409 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/39046 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/28377 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/81691 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/36853 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/28377 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28087 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4346 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/40805 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/42653 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/84367 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/42790 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/28377 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19053 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/WTF/wtf/RandomNumber.cpp:
```diff
@@ -41,6 +40,17 @@ double randomNumber()
+{
+    auto high = static_cast<uint64_t>(cryptographicallyRandomNumber());
+    return (high << 32) | cryptographicallyRandomNumber();
+}
```
Let's name it `cryptographicallyRandomUint64` etc. to emphasis it is cryptographically random.
We should not use it when cryptographically random numbers are required (if weak-random is enough, then use it).

donny-dont replied:
The convention in this file is "random" means cryptographically random while "weakRandom" means weak random. If I hadn't looked at the implementation I wouldn't have known that as the documentation doesn't imply that.

My opinion would be to keep the naming and then do an audit of where `random*` is used to figure out if they should be cryptographically random or not.

Constellation replied:
However we have real problem that, since `randomNumber` name looks so canonical and looks easy, many places were wrongly using that function. And the problem was that it is surprisingly costly function.
When working on Speedometer2.1 improvement, profiler trace showed many strong random generations, and it turned out most of them are OK with weak random. Not to repeat this mistake, we should name strong version of this well with the descriptive name to require code contributors to think about which one should be picked in that callsite.

mcatanzaro replied:
@donny-dont I would name the new functions as @Constellation suggests, even though it does not parallel the naming of the existing function, then report a follow-up bug to rename the existing function.

Constellation reviewed (approved):
r=me with comment.

donny-dont committed:
Add functions for generating random unsigned integers
https://bugs.webkit.org/show_bug.cgi?id=247446

Reviewed by Yusuke Suzuki.

Add `cryptographicallyRandomUint32` and `cryptographicallyRandomUint64`
for generating unsigned integers. Replace uses of `randomNumber` that
were creating random unsigned integers. These were all just reversing
the operations used internally in `randomNumber`.

* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
* Source/WTF/wtf/RandomNumber.cpp:
* Source/WTF/wtf/RandomNumber.h:
* Source/WebCore/platform/network/FormDataBuilder.cpp:
* Source/WebCore/style/RuleSetBuilder.cpp:
* Source/WebKit/Platform/IPC/win/ConnectionWin.cpp:
* Source/WebKit/Platform/unix/SharedMemoryUnix.cpp:

Canonical link: https://commits.webkit.org/256293@main

darinadler commented on Source/WebKit/Platform/IPC/win/ConnectionWin.cpp:
```diff
@@ -44,7 +44,7 @@ bool createServerAndClientIdentifiers(HANDLE& serverIdentifier, HANDLE& clientId
 
     do {
-        unsigned uniqueID = randomNumber() * std::numeric_limits<unsigned>::max();
+        unsigned uniqueID = cryptographicallyRandomUint32();
```
I would get rid of this local variable just as you did in `pushCascadeLayer`.

darinadler commented on Source/WebKit/Platform/IPC/win/ConnectionWin.cpp:
```diff
@@ -44,7 +44,7 @@ bool createServerAndClientIdentifiers(HANDLE& serverIdentifier, HANDLE& clientId
     do {
-        unsigned uniqueID = randomNumber() * std::numeric_limits<unsigned>::max();
+        unsigned uniqueID = cryptographicallyRandomUint32();
         pipeName = makeString("\\\\.\\pipe\\com.apple.WebKit.", hex(uniqueID));
```
So unclear on why we use hex here instead of just decimal.

darinadler commented on Source/WebCore/style/RuleSetBuilder.cpp:
```diff
@@ -224,8 +224,7 @@ void RuleSetBuilder::pushCascadeLayer(const CascadeLayerName& name)
             // Make unique name for an anonymous layer.
-            unsigned long long random = randomNumber() * std::numeric_limits<unsigned long long>::max();
-            return CascadeLayerName { makeAtomString("anon_"_s, random) };
+            return CascadeLayerName { makeAtomString("anon_"_s, cryptographicallyRandomUint64()) };
```
Do we really need 64 bits here?

darinadler commented on Source/WebKit/Platform/unix/SharedMemoryUnix.cpp:
```diff
@@ -140,7 +140,7 @@ static UnixFileDescriptor createSharedMemory()
     CString tempName;
     for (int tries = 0; fileDescriptor == -1 && tries < 10; ++tries) {
-        auto name = makeString("/WK2SharedMemory.", static_cast<unsigned>(WTF::randomNumber() * (std::numeric_limits<unsigned>::max() + 1.0)));
+        auto name = makeString("/WK2SharedMemory.", WTF::cryptographicallyRandomUint32());
```
Should not need the `WTF::` prefix here.

darinadler commented on Source/WTF/wtf/RandomNumber.h:
```diff
@@ -25,15 +25,25 @@
 // Returns a cryptographically secure pseudo-random number in the range [0, 1).
 WTF_EXPORT_PRIVATE double randomNumber();
 
+// Returns a cryptographically secure pseudo-random number in the range (0, UINT_MAX].
+WTF_EXPORT_PRIVATE unsigned cryptographicallyRandomUint32();
+
+// Returns a cryptographically secure pseudo-random number in the range (0, UINT64_MAX].
+WTF_EXPORT_PRIVATE uint64_t cryptographicallyRandomUint64();
```
Really glad you did this! I have been sick so missed my chance to review. I believe you already landed this, but:

- I would capitalize the "I" for "integer" and called them `cryptographicallyRandomUInt32` and `64`
- I would use the type `uint32_t` instead of `unsigned`, even though this is really unimportant and unclear when we should do that
- I would have stuck with the naming in this header and just called these `randomUInt32` and `randomUInt64`
- Or I would have renamed `randomNumber` to `cryptographicallyRandomDouble`, since it’s not so widely used we can’t rename
- I would also want to audit callers of `cryptographicallyRandomDouble` to see if they are really OK, since I am not sure it has enough random bits in it, not really a "random double" in my view

darinadler commented on Source/WTF/wtf/RandomNumber.cpp:
```diff
@@ -41,6 +40,17 @@ double randomNumber()
+
+uint64_t cryptographicallyRandomUint64()
+{
+    auto high = static_cast<uint64_t>(cryptographicallyRandomNumber());
```
I would just write:
```
    uint64_t high = cryptographicallyRandomNumber();
```
